### PR TITLE
Cubature Kalman Filter

### PR DESCRIPTION
**Pull request by jmbarr**
Provides a cubature Kalman predictor and updater. (Credit should go to Bethany? I think.) 

Side issue is that there's some odd behaviour at certain configurations. I'll provide a gist to highlight that.

### CODE REVIEW

File: stonesoup/functions/__init__.py [hunk 1]
```diff
@@ -59,8 +60,8 @@ def cholesky_eps(A, lower=False):
     L = np.zeros(A.shape)
     for i in range(A.shape[0]):
         for j in range(i):
-            L[i, j] = (A[i, j] - L[i, :]@L[j, :].T) / L[j, j]
-        val = A[i, i] - L[i, :]@L[i, :].T
+            L[i, j] = (A[i, j] - L[i, :] @ L[j, :].T) / L[j, j]
+        val = A[i, i] - L[i, :] @ L[i, :].T
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:64]: This was correct before.

File: stonesoup/functions/__init__.py [hunk 2]
```diff
@@ -91,19 +92,20 @@ def jacobian(fun, x,  **kwargs):
 
     # For numerical reasons the step size needs to large enough. Aim for 1e-8
     # relative to spacing between floating point numbers for each dimension
-    delta = 1e8*np.spacing(x.state_vector.astype(np.float64).ravel())
+    delta = 1e8 * np.spacing(x.state_vector.astype(np.float_).ravel())
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:95]: Should use `np.float64` still

File: stonesoup/functions/__init__.py [hunk 3]
```diff
@@ -91,19 +92,20 @@ def jacobian(fun, x,  **kwargs):
...
     # But at least 1e-8
     # TODO: Is this needed? If not, note special case at zero.
     delta[delta < 1e-8] = 1e-8
 
     x2 = copy.copy(x)  # Create a clone of the input
-    x2.state_vector = np.tile(x.state_vector, ndim+1) + np.eye(ndim, ndim+1)*delta[:, np.newaxis]
+    x2.state_vector = np.tile(x.state_vector, ndim + 1) + np.eye(ndim, ndim + 1) * \
+        delta[:, np.newaxis]
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:102]: Line wasn't too long before

File: stonesoup/functions/__init__.py [hunk 4]
```diff
@@ -91,19 +92,20 @@ def jacobian(fun, x,  **kwargs):
...
     x2.state_vector = x2.state_vector.view(StateVectors)
 
     F = fun(x2, **kwargs)
 
     jac = np.divide(F[:, :ndim] - F[:, -1:], delta)
-    return jac.astype(np.float64)
+    return jac.astype(np.float_)
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:108]: Correct before

File: stonesoup/functions/__init__.py [hunk 5]
```diff
@@ -165,9 +167,9 @@ def gauss2sigma(state, alpha=1.0, beta=2.0, kappa=None):
 
     # Can't use in place addition/subtraction as casting issues may arise when mixing float/int
     sigma_points[:, 1:(ndim_state + 1)] = \
-        sigma_points[:, 1:(ndim_state + 1)] + sqrt_sigma*np.sqrt(c)
+        sigma_points[:, 1:(ndim_state + 1)] + sqrt_sigma * np.sqrt(c)
     sigma_points[:, (ndim_state + 1):] = \
-        sigma_points[:, (ndim_state + 1):] - sqrt_sigma*np.sqrt(c)
+        sigma_points[:, (ndim_state + 1):] - sqrt_sigma * np.sqrt(c)
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:172]: This was correct before

File: stonesoup/functions/__init__.py [hunk 6]
```diff
@@ -280,7 +282,8 @@ def unscented_transform(sigma_points_states, mean_weights, covar_weights,
 
     # Calculate cross-covariance
     cross_covar = (
-        (sigma_points-sigma_points[:, 0:1]) @ np.diag(mean_weights) @ (sigma_points_t-mean).T
+            (sigma_points - sigma_points[:, 0:1]) @ np.diag(mean_weights) @
+            (sigma_points_t - mean).T
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:286]: This was correct before.

File: stonesoup/functions/__init__.py [hunk 7]
```diff
@@ -303,7 +306,7 @@ def cart2pol(x, y):
 
     """
 
-    rho = np.sqrt(x**2 + y**2)
+    rho = np.sqrt(x ** 2 + y ** 2)
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:309]: Correct before

File: stonesoup/functions/__init__.py [hunk 8]
```diff
@@ -328,9 +331,9 @@ def cart2sphere(x, y, z):
 
     """
 
-    rho = np.sqrt(x**2 + y**2 + z**2)
+    rho = np.sqrt(x ** 2 + y ** 2 + z ** 2)
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:334]: Correct before

File: stonesoup/functions/__init__.py [hunk 9]
```diff
@@ -419,7 +422,7 @@ def cart2az_el_rg(x, y, z):
     (float, float, float)
         A tuple of the form `(phi, theta, rho)`
     """
-    rho = np.sqrt(x**2 + y**2 + z**2)
+    rho = np.sqrt(x ** 2 + y ** 2 + z ** 2)
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:425]: Correct before

File: stonesoup/functions/__init__.py [hunk 10]
```diff
@@ -626,7 +629,8 @@ def gm_reduce_single(means, covars, weights):
 
     # Calculate covar
     delta_means = means - mean
-    covar = np.sum(covars*weights, axis=2, dtype=np.float64) + weights*delta_means@delta_means.T
+    covar = np.sum(covars * weights, axis=2, dtype=np.float_) + \
+        weights * delta_means @ delta_means.T
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:633]: Line wasn't too long before

File: stonesoup/functions/__init__.py [hunk 11]
```diff
@@ -759,5 +763,163 @@ def sde_euler_maruyama_integration(fun, t_values, state_x0):
         delta_t = next_t - t
         delta_w = np.random.normal(scale=np.sqrt(delta_t), size=(state_x.ndim, 1))
         a, b = fun(state_x, t)
-        state_x.state_vector = state_x.state_vector + a*delta_t + b@delta_w
+        state_x.state_vector = state_x.state_vector + a * delta_t + b @ delta_w
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:766]: Correct before

File: stonesoup/functions/__init__.py [hunk 12]
```diff
@@ -759,5 +763,163 @@ def sde_euler_maruyama_integration(fun, t_values, state_x0):
...
+    ndim_state = np.shape(state.state_vector)[0]
+
+    sqrt_covar = np.linalg.cholesky(state.covar)
+    cuba_points = np.sqrt(alpha*ndim_state) * np.hstack((np.identity(ndim_state),
+                                                         -np.identity(ndim_state)))
+
+    if np.issubdtype(cuba_points.dtype, np.integer):
+        cuba_points = cuba_points.astype(float)
+
+    cuba_points = sqrt_covar @ cuba_points + state.mean
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:810]:
```suggestion
    cuba_points = sqrt_covar@cuba_points + state.mean
```

File: stonesoup/functions/__init__.py [hunk 13]
```diff
@@ -759,5 +763,163 @@ def sde_euler_maruyama_integration(fun, t_values, state_x0):
...
+    """
+    ndim_state = np.shape(state.state_vector)[0]
+    cubature_points = gauss2cubature(state)
+
+    if points_noise is None:
+        cubature_points_t = StateVectors([fun(State(cub_point)) for cub_point in cubature_points])
+    else:
+        cubature_points_t = StateVectors(
+            [fun(State(cub_point), points_noise) for cub_point, point_noise in
+             zip(cubature_points, points_noise)])
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:917]:
```suggestion
        cubature_points_t = StateVectors([
            fun(State(cub_point), points_noise)
            for cub_point, point_noise in zip(cubature_points, points_noise)])
```

File: stonesoup/functions/__init__.py [hunk 14]
```diff
@@ -759,5 +763,163 @@ def sde_euler_maruyama_integration(fun, t_values, state_x0):
...
+
+    mean, covar = cubature2gauss(cubature_points_t, covar_noise)
+
+    cross_covar = (1/alpha)*((1.0 / (2 * ndim_state)) * cubature_points @ cubature_points_t.T -
+                             np.average(cubature_points, axis=1) @ mean.T)
```
Comment (sdhiscocks) [inline stonesoup/functions/__init__.py:922]:
```suggestion
    cross_covar = (1/alpha)*((1.0 / (2 * ndim_state)) * cubature_points @ cubature_points_t.T
                             - np.average(cubature_points, axis=1) @ mean.T)
```

File: stonesoup/predictor/kalman.py [hunk 15]
```diff
@@ -474,3 +474,95 @@ def _predicted_covariance(self, prior, predict_over_interval, control_input=None
...
+            "will create a zero-effect linear :class:`~.ControlModel`.")
+    alpha: float = Property(
+        default=1.0,
+        doc="Scaling parameter. Default is 1.0. Lower values select points closer to the mean and "
+            "vice versa.")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._time_interval = None
```
Comment (sdhiscocks) [inline stonesoup/predictor/kalman.py:501]:
```suggestion
```

File: stonesoup/predictor/kalman.py [hunk 16]
```diff
@@ -474,3 +474,95 @@ def _predicted_covariance(self, prior, predict_over_interval, control_input=None
...
+
+        Returns
+        -------
+        : :class:`numpy.ndarray`
+            The combined, noiseless, effect of applying the transition and
+            control
+        """
+
+        return self.transition_model.function(prior_state, **kwargs) + \
+            self.control_model.function(**kwargs)
```
Comment (sdhiscocks) [inline stonesoup/predictor/kalman.py:522]:
```suggestion
        return self.transition_model.function(prior_state, **kwargs) \
            + self.control_model.function(**kwargs)
```

File: stonesoup/predictor/kalman.py [hunk 17]
```diff
@@ -474,3 +474,95 @@ def _predicted_covariance(self, prior, predict_over_interval, control_input=None
...
+        transition_and_control_function = partial(
+            self._transition_and_control_function,
+            control_input=control_input,
+            time_interval=predict_over_interval)
+
+        x_pred, p_pred, _, _ = cubature_transform(prior, transition_and_control_function,
+                                                  covar_noise=total_noise_covar, alpha=self.alpha)
+
+        return Prediction.from_state(prior, x_pred, p_pred, timestamp=timestamp,
+                                     transition_model=self.transition_model)
```
Comment (sdhiscocks) [inline stonesoup/predictor/kalman.py:568]:
```suggestion
                                     transition_model=self.transition_model, prior=prior)
```

**Review by sdhiscocks — APPROVED** [review]
Looks good 👍  Just some minor formatting changes, undo changes that'll break NumPy 2.0 compatibility, and new additional `prior` for Prediction type.

**Comment by jmbarr** [comment]
> Looks good 👍 Just some minor formatting changes, undo changes that'll break NumPy 2.0 compatibility, and new additional `prior` for Prediction type.

Don't think I did any of these formatting changes and can't explain how they got in. Is there an 'undo' button?

**Review by jswright-dstl — APPROVED** [review]

**Pull request merged**